# The Fermat quartic surface as a Kummer surface

The claim to be pinned down: the diagonal quartic $x^4+y^4+z^4=1$ — projectively
$x_0^4+x_1^4+x_2^4+x_3^4=0$ over $\mathbb{Q}(\mu_8)$, since $\zeta_8^4=-1$ — is isomorphic to a
Kummer surface, the isomorphism existing only over $\mathbb{Q}(\mu_8) = \mathbb{Q}(\sqrt2, i)$.

This is **Mizukami's theorem** (Master's thesis, Tokyo 1977). Ieronymou–Skorobogatov–Zarhin use it as
the engine of *On the Brauer group of diagonal quartic surfaces* ([arXiv:0912.2865](https://arxiv.org/abs/0912.2865));
it is their Theorem A.1, and Sir Peter Swinnerton-Dyer's appendix to that paper reconstructs the proof
and writes the map down. Everything below is a verification of that appendix, plus a simplification of
its formulas.

**One correction to the folklore version.** It is tempting to remember the statement as
$X \cong \operatorname{Kum}(E\times E)$ with $E$ the lemniscatic curve, $j(E)=1728$. That is not quite
right, and ISZ say so explicitly: *"There is a rational isogeny $A \to E\times E$ of degree $2$, where
$E$ is the elliptic curve $y^2=x^3-4x$."* The abelian surface is only **$2$-isogenous** to $E\times E$.
Section 2 identifies it exactly: $A \cong E_i \times E_{2i}$, a product of two *non-isomorphic* CM
curves, with $j$-invariants $1728$ and $287496$. The distinction is not cosmetic —
$\operatorname{Kum}(E_i\times E_i)$ is a different K3 surface, with a different transcendental lattice.

## Notation

Throughout $\epsilon = \zeta_8$, so $\epsilon^4 = -1$, $i = \epsilon^2$, $\sqrt2 = \epsilon+\epsilon^7$.
$X \subset \mathbb{P}^3$ is the Fermat quartic $F = x_0^4+x_1^4+x_2^4+x_3^4 = 0$.

## 1. The curve $C$ and the abelian surface $A$

Swinnerton-Dyer starts from the genus-one curve
$$C: \quad v^2 = (u^2-1)\left(u^2-\tfrac12\right) = u^4 - \tfrac32 u^2 + \tfrac12,$$
with base point $O$ the point at infinity where $v/u^2 = 1$, and $T$ the *other* point at infinity,
where $v/u^2 = -1$.

$T$ is a $2$-division point. Indeed, with $c = \tfrac34$ one computes
$v^2 - (u^2-c)^2 = -\tfrac1{16}$, a nonzero constant, so $\operatorname{div}(v - u^2 + \tfrac34) = 2O - 2T$
and $2(O-T) = 0$. The involution $\tau: (u,v)\mapsto(-u,-v)$ is translation by $T$, hence fixed point
free, and
$$A \;=\; (C\times C)\big/\{(O,O),\,(T,T)\}$$
is an abelian surface, with $C\times C \to A$ of degree $2$.

To put $C$ in Weierstrass form, write $s = v+u^2$; then $u^2(2s+a) = s^2-b$ for $v^2 = u^4+au^2+b$, and
multiplying up gives
$$Y^2 = (X+a)(X^2-4b), \qquad X = 2(v+u^2), \quad Y = 2u\bigl(2(v+u^2)+a\bigr).$$
Here $a = -\tfrac32$, $b = \tfrac12$, so $Y^2 = (X-\tfrac32)(X^2-2)$. The point $T$ has $s = \tfrac34$,
i.e. $X = \tfrac32$, $Y=0$ — it is the rational $2$-torsion point.

In [1]:
# C in Weierstrass form:  Y^2 = (X - 3/2)(X^2 - 2) = X^3 - 3/2 X^2 - 2X + 3
C = EllipticCurve(QQ, [0, -3/2, 0, -2, 3]).minimal_model()
print("C        :", C)
print("           label", C.cremona_label(),
      "  j =", C.j_invariant(), "=", factor(ZZ(C.j_invariant())),
      "  CM by disc", C.cm_discriminant())
print("C[2](Q)  : x-coordinates", C.division_polynomial(2).roots(QQ, multiplicities=False),
      " -> T is rational, generating ker(C -> C/tau)")

for ph in C.isogenies_prime_degree(2):
    Cp = ph.codomain().minimal_model()
    print("C' = C/tau:", Cp)
    print("           label", Cp.cremona_label(),
          "  j =", Cp.j_invariant(),
          "  CM by disc", Cp.cm_discriminant())

C        : Elliptic Curve defined by y^2 = x^3 - 44*x + 112 over Rational Field


           label 64a3   j = 287496 = 2^3 * 3^3 * 11^3   CM by disc -16
C[2](Q)  : x-coordinates [4]  -> T is rational, generating ker(C -> C/tau)
C' = C/tau: Elliptic Curve defined by y^2 = x^3 - 4*x over Rational Field
           label 64a1   j = 1728   CM by disc -4


So $C$ has $j = 287496 = 66^3$ and CM by $\mathbb{Z}[2i]$ (discriminant $-16$), while the quotient
$C' = C/\tau$ has $j = 1728$ and CM by $\mathbb{Z}[i]$ (discriminant $-4$). In Cremona's tables
$C' = \texttt{64a1}$, the curve $y^2 = x^3-4x$ that ISZ name — and over $\mathbb{Q}(\sqrt2)$ that is the
lemniscatic curve $y^2=x^3-x$.

(Swinnerton-Dyer labels $C$ as `64a2`; Sage's minimal model of my ad hoc Weierstrass equation lands on
`64a3`, the other quadratic twist. Same $j$, same CM order, immaterial here.)

In [2]:
K2.<s2> = QuadraticField(2)
print("y^2 = x^3 - 4x  ==  y^2 = x^3 - x  over Q(sqrt2) ?",
      EllipticCurve(K2, [0,0,0,-4,0]).is_isomorphic(EllipticCurve(K2, [0,0,0,-1,0])))

y^2 = x^3 - 4x  ==  y^2 = x^3 - x  over Q(sqrt2) ? True


## 2. $A \cong E_i \times E_{2i}$ — and why that is not $E\times E$

$A$ splits, and one can see it in a line. Let $\pi: C \to C' = C/\tau$ be the quotient isogeny, so
$\ker\pi = \{O,T\}$. Consider the homomorphism of abelian surfaces
$$\Phi: C\times C \longrightarrow C'\times C, \qquad \Phi(P,Q) = \bigl(\pi(P),\; Q-P\bigr).$$
Then $\Phi(P,Q) = 0$ forces $P \in \{O,T\}$ and $Q = P$, so
$$\ker\Phi = \{(O,O),\,(T,T)\},$$
which is exactly the subgroup we quotiented by. Hence $\Phi$ descends to an isomorphism
$$\boxed{\;A \;\cong\; C' \times C \;\cong\; E_i \times E_{2i}\;}$$
with $j$-invariants $1728$ and $287496$ respectively. Since $j(C) \neq j(C')$, the two factors are not
isomorphic, so $A \not\cong E\times E$ for any single $E$. The degree-$2$ isogeny $A \to E\times E$ that
ISZ mention is just $\mathrm{id} \times \pi$ applied the other way: $E_{2i}$ is $2$-isogenous to $E_i$.

**Cross-check by lattices.** This is a consistency check, not an independent proof: it takes
$T(X) = \operatorname{diag}(8,8)$ as a known input (Piatetskii-Shapiro–Shafarevich, Shioda–Inose), which
is *not* verified in this notebook. The isogeny argument just given is self-contained and needs none of it.

Shioda–Mitani classify singular abelian surfaces by positive definite even
binary quadratic forms, $A_{[a,b,c]} = E_{\tau_1}\times E_{\tau_2}$ with $\tau_1 = \frac{-b+\sqrt D}{2a}$,
$\tau_2 = \frac{b+\sqrt D}{2}$, $D = b^2-4ac$. And for any abelian surface,
$T(\operatorname{Kum}(A)) \cong T(A)(2)$. The Fermat quartic has $T(X) = \operatorname{diag}(8,8)$, so
$T(A) = \operatorname{diag}(4,4)$, the form $[2,0,2]$ of discriminant $-16$, giving
$\tau_1 = i$, $\tau_2 = 2i$: exactly $E_i\times E_{2i}$. Whereas $E_i\times E_i$ is the form $[1,0,1]$,
$T = \operatorname{diag}(2,2)$, and $\operatorname{Kum}(E_i\times E_i)$ has
$T = \operatorname{diag}(4,4) \neq \operatorname{diag}(8,8)$ — a genuinely different K3 surface.

Note that point counting cannot see this distinction: $E_i\times E_{2i}$ and $E_i\times E_i$ are
isogenous, so their Kummer surfaces have identical zeta functions. Only the lattice — or the explicit
map of Section 6 — settles it.

In [3]:
print("Shioda-Mitani bookkeeping")
print("  A            T(A)                T(Kum A)")
for name, T in [("E_i  x E_i ", matrix(2,2,[2,0,0,2])), ("E_i  x E_2i", matrix(2,2,[4,0,0,4]))]:
    print("  %s  diag(%d,%d)  disc %-4d   diag(%d,%d)  disc %d"
          % (name, T[0,0], T[1,1], -T.det(), 2*T[0,0], 2*T[1,1], -(2*T).det()))
print()
print("  Fermat quartic has T(X) = diag(8,8), disc -64  ->  T(A) = diag(4,4)  ->  A = E_i x E_2i")

Shioda-Mitani bookkeeping
  A            T(A)                T(Kum A)
  E_i  x E_i   diag(2,2)  disc -4     diag(4,4)  disc -16
  E_i  x E_2i  diag(4,4)  disc -16    diag(8,8)  disc -64

  Fermat quartic has T(X) = diag(8,8), disc -64  ->  T(A) = diag(4,4)  ->  A = E_i x E_2i


## 3. The Kummer surface $K$ in coordinates

The function field of $C\times C$ is generated by $u_1,v_1,u_2,v_2$. The involution $P \mapsto -P$ on
$A$ reverses the signs of $u_1,u_2$ (and of $v_1,v_2$ together with the $\tau$-shift), so $k(K)$ is the
field of functions even in all four variables collectively. Swinnerton-Dyer takes
$$w_1 = u_1^2, \qquad w_2 = u_2^2, \qquad y = \frac{v_1(w_2-1)}{v_2\bigl(w_1-\tfrac12\bigr)},
\qquad z = \frac{u_2}{u_1},$$
so that, up to birational equivalence, $K = \operatorname{Kum}(A)$ is
$$y^2 \;=\; \frac{(w_1-1)(w_2-1)}{\bigl(w_1-\tfrac12\bigr)\bigl(w_2-\tfrac12\bigr)},
\qquad\qquad z^2 \;=\; \frac{w_2}{w_1}. \tag{6}$$
In particular $[k(K):k(w_1,w_2)] = 4$: the plane of the $w_j$ is the quotient of $C\times C$ by
$\{\pm1\}\times\{\pm1\}$, and $K$ sits halfway.

The first equation is just $v_1^2 v_2^{-2}$ rewritten using $v_j^2 = (w_j-1)(w_j-\tfrac12)$; the odd
looking asymmetry in the definition of $y$ is engineered so that the formula in Section 6 comes out clean.

## 4. Quadrics on the Fermat quartic

Swinnerton-Dyer's notation. The $48$ lines on $X$ are $L_{\mu\nu}: x_0 = \epsilon^\mu x_1,\, x_2 = \epsilon^\nu x_3$
and the two other pairings $M_{\mu\nu}$, $N_{\mu\nu}$, for $\mu,\nu$ odd mod $8$. The quadrics that matter are
$$e'_\pm = x_0x_3 \pm x_1x_2, \qquad e''_\pm = x_0x_2 \pm x_1x_3,$$
$$h'_{\alpha\beta} = x_0^2 - \epsilon^{2\alpha}x_1^2 - \epsilon^{2\beta}x_2^2 + \epsilon^{2\alpha+2\beta}x_3^2,
\qquad
h''_{\alpha\beta} = x_0^2 - \epsilon^{2\alpha}x_1^2 - \epsilon^{2\beta}x_3^2 + \epsilon^{2\alpha+2\beta}x_2^2,$$
together with the linear forms $f_{\lambda\mu\nu} = x_0 + \epsilon^\lambda x_1 + \epsilon^\mu x_2 + \epsilon^\nu x_3$.

The identities the whole construction turns on are his (7):
$$h'_{13}h'_{31} \;=\; -h'_{11}h'_{33} \;=\; 2\,e'_+e'_-, \qquad
h''_{13}h''_{31} \;=\; -h''_{11}h''_{33} \;=\; 2\,e''_+e''_- \qquad \text{on } X. \tag{7}$$
These hold modulo $F$, not identically.

In [4]:
K.<ep> = CyclotomicField(8)          # ep = zeta_8
iu = ep^2                            # i
rt2 = ep + ep^7                      # sqrt(2)
assert ep^4 == -1 and iu^2 == -1 and rt2^2 == 2

R.<x0,x1,x2,x3> = PolynomialRing(K)
F = x0^4 + x1^4 + x2^4 + x3^4

def f(l, m, n):  return x0 + ep^(l % 8)*x1 + ep^(m % 8)*x2 + ep^(n % 8)*x3
def hp(a, b):    return x0^2 - ep^((2*a) % 8)*x1^2 - ep^((2*b) % 8)*x2^2 + ep^((2*a+2*b) % 8)*x3^2
def hpp(a, b):   return x0^2 - ep^((2*a) % 8)*x1^2 - ep^((2*b) % 8)*x3^2 + ep^((2*a+2*b) % 8)*x2^2

ep_p,  em_p  = x0*x3 + x1*x2, x0*x3 - x1*x2      # e'_+ , e'_-
ep_pp, em_pp = x0*x2 + x1*x3, x0*x2 - x1*x3      # e''_+, e''_-

print("identities (7), as congruences modulo F:")
print("   h'13 h'31  = -h'11 h'33  :", F.divides(hp(1,3)*hp(3,1) + hp(1,1)*hp(3,3)))
print("   h'13 h'31  = 2 e'+ e'-   :", F.divides(hp(1,3)*hp(3,1) - 2*ep_p*em_p))
print("   h''13 h''31 = -h''11 h''33:", F.divides(hpp(1,3)*hpp(3,1) + hpp(1,1)*hpp(3,3)))
print("   h''13 h''31 = 2 e''+ e''- :", F.divides(hpp(1,3)*hpp(3,1) - 2*ep_pp*em_pp))
print()
print("and none holds identically in P^3 -- the discrepancy is exactly F itself:")
print("   h'13 h'31  - 2 e'+ e'-  =", hp(1,3)*hp(3,1) - 2*ep_p*em_p)
print("   h''13 h''31 - 2 e''+ e''- =", hpp(1,3)*hpp(3,1) - 2*ep_pp*em_pp)

identities (7), as congruences modulo F:
   h'13 h'31  = -h'11 h'33  : True
   h'13 h'31  = 2 e'+ e'-   : True


   h''13 h''31 = -h''11 h''33: True
   h''13 h''31 = 2 e''+ e''- : True

and none holds identically in P^3 -- the discrepancy is exactly F itself:
   h'13 h'31  - 2 e'+ e'-  = x0^4 + x1^4 + x2^4 + x3^4
   h''13 h''31 - 2 e''+ e''- = x0^4 + x1^4 + x2^4 + x3^4


## 5. Collapsing Swinnerton-Dyer's $F_i$, $G_i$

The appendix produces $w_1,w_2$ as ratios of four functions $F_1,F_2,G_1,G_2$, each a quotient of a
product of six linear and two quadratic forms:
$$F_1 = \frac{f_{727}f_{763}f_{125}f_{161}\,e'_+\,(x_0-\epsilon x_1)(x_0-\epsilon^7x_1)}
             {e''_+(x_2-\epsilon x_3)(x_2-\epsilon^7x_3)},
\qquad
F_2 = \frac{f_{727}f_{763}f_{327}f_{363}\,h'_{13}\,(x_2-\epsilon^5x_3)}{h''_{33}(x_2-\epsilon x_3)},$$
$$G_1 = \frac{f_{367}f_{323}f_{125}f_{161}\,e'_-\,(x_0-\epsilon x_1)(x_0-\epsilon^3x_1)}
             {e''_+(x_2-\epsilon^5x_3)(x_2-\epsilon^7x_3)},
\qquad
G_2 = \frac{f_{367}f_{323}f_{327}f_{363}\,h'_{31}\,(x_0-\epsilon^3x_1)}{h''_{33}(x_0-\epsilon^7x_1)},$$
with $w_1 = \frac{\epsilon}{1+\epsilon^2}\frac{F_2}{F_1}$, $w_2 = \frac{\epsilon}{1+\epsilon^2}\frac{G_2}{G_1}$.

Almost all of this cancels. Write $P_\alpha = x_0+\epsilon^\alpha x_1$ and $Q_\beta = x_2+\epsilon^\beta x_3$.
Because $\epsilon^2+\epsilon^6 = \epsilon^3+\epsilon^7 = 0$, each of the four products
$f_{\lambda\mu\nu}f_{\lambda\mu'\nu'}$ occurring above has vanishing cross term and collapses to a sum
of two squares:
$$f_{727}f_{763} = P_7^2+Q_5^2,\quad f_{327}f_{363} = P_3^2+Q_5^2,\quad
f_{125}f_{161} = P_1^2+Q_3^2,\quad f_{367}f_{323} = P_3^2+Q_1^2 .$$
In $F_2/F_1$ the pair $f_{727}f_{763}$ cancels outright, in $G_2/G_1$ the pair $f_{367}f_{323}$ does, and
what survives is a *single common factor*
$$\Theta \;=\; \frac{(P_3^2+Q_5^2)\;e''_+\;Q_1Q_3}{(P_1^2+Q_3^2)\;h''_{33}\;P_5P_3},
\qquad\quad
\frac{F_2}{F_1} = \Theta\cdot\frac{h'_{13}}{e'_+}, \qquad
\frac{G_2}{G_1} = \Theta\cdot\frac{h'_{31}}{e'_-}.$$
Finally $1+\epsilon^2 = 1+i = \sqrt2\,\epsilon$, so the constant $\dfrac{\epsilon}{1+\epsilon^2}$ is simply
$\dfrac{1}{\sqrt2}$. This drops the bidegree of $w_1,w_2$ from $11/11$ to $8/8$.

In [5]:
P = lambda a: x0 + ep^(a % 8)*x1
Q = lambda b: x2 + ep^(b % 8)*x3

print("the four f-products are sums of two squares:")
print("   f727 f763 = P7^2 + Q5^2 :", f(7,2,7)*f(7,6,3) == P(7)^2 + Q(5)^2)
print("   f327 f363 = P3^2 + Q5^2 :", f(3,2,7)*f(3,6,3) == P(3)^2 + Q(5)^2)
print("   f125 f161 = P1^2 + Q3^2 :", f(1,2,5)*f(1,6,1) == P(1)^2 + Q(3)^2)
print("   f367 f323 = P3^2 + Q1^2 :", f(3,6,7)*f(3,2,3) == P(3)^2 + Q(1)^2)
print()
print("   eps/(1+eps^2) == 1/sqrt2 :", ep/(1+ep^2) == 1/rt2)
print()

# Swinnerton-Dyer's F1, F2, G1, G2 as (numerator, denominator)
F1n = f(7,2,7)*f(7,6,3)*f(1,2,5)*f(1,6,1)*ep_p*(x0-ep*x1)*(x0-ep^7*x1)
F1d = ep_pp*(x2-ep*x3)*(x2-ep^7*x3)
F2n = f(7,2,7)*f(7,6,3)*f(3,2,7)*f(3,6,3)*hp(1,3)*(x2-ep^5*x3)
F2d = hpp(3,3)*(x2-ep*x3)
G1n = f(3,6,7)*f(3,2,3)*f(1,2,5)*f(1,6,1)*em_p*(x0-ep*x1)*(x0-ep^3*x1)
G1d = ep_pp*(x2-ep^5*x3)*(x2-ep^7*x3)
G2n = f(3,6,7)*f(3,2,3)*f(3,2,7)*f(3,6,3)*hp(3,1)*(x0-ep^3*x1)
G2d = hpp(3,3)*(x0-ep^7*x1)

# Theta
Thn = (P(3)^2 + Q(5)^2) * ep_pp   * Q(1)*Q(3)
Thd = (P(1)^2 + Q(3)^2) * hpp(3,3)* P(5)*P(3)

print("the cancellations are exact identities in the polynomial ring, no reduction mod F needed:")
print("   F2/F1 == Theta * h'13/e'+ :", F2n*F1d*Thd*ep_p == Thn*hp(1,3)*F2d*F1n)
print("   G2/G1 == Theta * h'31/e'- :", G2n*G1d*Thd*em_p == Thn*hp(3,1)*G2d*G1n)

the four f-products are sums of two squares:
   f727 f763 = P7^2 + Q5^2 : True
   f327 f363 = P3^2 + Q5^2 : True
   f125 f161 = P1^2 + Q3^2 : True
   f367 f323 = P3^2 + Q1^2 : True

   eps/(1+eps^2) == 1/sqrt2 : True

the cancellations are exact identities in the polynomial ring, no reduction mod F needed:
   F2/F1 == Theta * h'13/e'+ : True
   G2/G1 == Theta * h'31/e'- : True


## 6. The map, in closed form

The last simplification uses (7) once more. With $z = \epsilon^3(1+\epsilon^2)\dfrac{e'_+}{h'_{13}}$,
equation (7) gives $\dfrac{h'_{31}}{e'_-} = \dfrac{2e'_+}{h'_{13}}$ on $X$, and
$\epsilon^3(1+\epsilon^2) = \epsilon^3\sqrt2\,\epsilon = -\sqrt2$, so the whole map becomes

$$\boxed{\;
z = -\sqrt2\,\frac{x_0x_3+x_1x_2}{h'_{13}}, \qquad
y = 2i\,\frac{x_0x_2+x_1x_3}{h''_{11}}, \qquad
w_1 = -\frac{\Theta}{z}, \qquad w_2 = -\Theta z \;}$$

with
$$h'_{13} = x_0^2 - ix_1^2 + ix_2^2 + x_3^2, \qquad h''_{11} = x_0^2 - ix_1^2 - x_2^2 - ix_3^2,$$
$$\Theta = \frac{\bigl[(x_0+\epsilon^3x_1)^2+(x_2-\epsilon x_3)^2\bigr]\;(x_0x_2+x_1x_3)\;
                 \bigl(x_2^2 + i\sqrt2\,x_2x_3 - x_3^2\bigr)}
                {\bigl[(x_0+\epsilon x_1)^2+(x_2+\epsilon^3x_3)^2\bigr]\;
                 \bigl(x_0^2+ix_1^2-x_2^2+ix_3^2\bigr)\;
                 \bigl(x_0^2-\sqrt2\,x_0x_1+x_1^2\bigr)} .$$

Note $w_1w_2 = \Theta^2$ and $w_2/w_1 = z^2$: the second equation of (6) is built in, and $\Theta$ is
the "geometric mean" of the two elliptic pencil parameters. The cell below checks both equations of (6)
as exact ideal memberships over $\mathbb{Q}(\zeta_8)$ — no numerics.

In [6]:
# everything written out, independent of the cells above except the ring
ep_p  = x0*x3 + x1*x2                          # e'_+
ep_pp = x0*x2 + x1*x3                          # e''_+
h13   = x0^2 - iu*x1^2 + iu*x2^2 +    x3^2     # h'_13
h11   = x0^2 - iu*x1^2 -    x2^2 - iu*x3^2     # h''_11
h33   = x0^2 + iu*x1^2 -    x2^2 + iu*x3^2     # h''_33
N1    = (x0 + ep^3*x1)^2 + (x2 - ep*x3)^2      # f327 f363
D1    = (x0 + ep  *x1)^2 + (x2 + ep^3*x3)^2    # f125 f161
N2    = x2^2 + iu*rt2*x2*x3 - x3^2             # Q1 Q3
D2    = x0^2 - rt2*x0*x1    + x1^2             # P5 P3

Thn, Thd = N1*ep_pp*N2, D1*h33*D2               # Theta
zn,  zd  = -rt2*ep_p, h13                       # z
yn,  yd  = 2*iu*ep_pp, h11                      # y
w1n, w1d = -Thn*zd, Thd*zn                      # w1 = -Theta/z
w2n, w2d = -Thn*zn, Thd*zd                      # w2 = -Theta*z

A1, A2 = w1n - w1d,   w2n - w2d                 # w_j - 1
B1, B2 = w1n - w1d/2, w2n - w2d/2               # w_j - 1/2

print("equations (6), as congruences modulo F:")
print("   z^2 = w2/w1                             :",
      F.divides(zn^2*w1n*w2d - zd^2*w2n*w1d))
print("   y^2 = (w1-1)(w2-1)/((w1-1/2)(w2-1/2))   :",
      F.divides(yn^2*B1*B2 - yd^2*A1*A2))
print()
print("   w1*w2 = Theta^2                          :",
      F.divides(w1n*w2n*Thd^2 - Thn^2*w1d*w2d))
print()
print("bidegrees:  Theta %d/%d   w1 %d/%d   w2 %d/%d   y %d/%d   z %d/%d"
      % (Thn.degree(), Thd.degree(), w1n.degree(), w1d.degree(),
         w2n.degree(), w2d.degree(), yn.degree(), yd.degree(), zn.degree(), zd.degree()))

equations (6), as congruences modulo F:
   z^2 = w2/w1                             : True
   y^2 = (w1-1)(w2-1)/((w1-1/2)(w2-1/2))   : True

   w1*w2 = Theta^2                          : True

bidegrees:  Theta 6/6   w1 8/8   w2 8/8   y 2/2   z 2/2


## 7. Birationality, and a point count

Swinnerton-Dyer argues that $w_1 = \text{const}$ and $w_2 = \text{const}$ are members of the two pencils
$\mathcal{P}'$, $\mathcal{P}''$ whose intersection number is $4$, so $[k(X):k(w_1,w_2)] = 4 = [k(K):k(w_1,w_2)]$
and the map is birational. Since $X$ and $K$ are both minimal models — smooth K3 surfaces — a birational
map between them is biregular, and that is the theorem.

Rather than re-run the divisor-theoretic argument, here is a direct check over $\mathbb{F}_{41}$
(chosen because $41\equiv 1 \bmod 8$, so $\zeta_8 \in \mathbb{F}_{41}$ and everything is defined there).
We enumerate $X(\mathbb{F}_{41})$ and confirm that $P \mapsto (w_1,w_2,y,z)$ is *injective* where it is
defined, while $P\mapsto (w_1,w_2)$ is generically $4:1$.

In [7]:
p = 41
Fp = GF(p)
E8 = [t for t in Fp if t and t.multiplicative_order() == 8][0]
assert E8^4 == -1
Rp.<y0,y1,y2,y3> = PolynomialRing(Fp)
IU, RT2 = E8^2, E8 + E8^7

Pp = lambda a: y0 + E8^(a % 8)*y1
Qp = lambda b: y2 + E8^(b % 8)*y3
epP  = y0*y3 + y1*y2
epPP = y0*y2 + y1*y3
H13  = y0^2 - IU*y1^2 + IU*y2^2 +    y3^2
H11  = y0^2 - IU*y1^2 -    y2^2 - IU*y3^2
H33  = y0^2 + IU*y1^2 -    y2^2 + IU*y3^2
THn  = ((y0+E8^3*y1)^2 + (y2-E8*y3)^2) * epPP * (y2^2 + IU*RT2*y2*y3 - y3^2)
THd  = ((y0+E8  *y1)^2 + (y2+E8^3*y3)^2) * H33 * (y0^2 - RT2*y0*y1 + y1^2)
Zn, Zd = -RT2*epP, H13
Yn, Yd = 2*IU*epPP, H11
MAPS = {'w1': (-THn*Zd, THd*Zn), 'w2': (-THn*Zn, THd*Zd), 'y': (Yn, Yd), 'z': (Zn, Zd)}

def P3points():
    out = [(Fp(1), a, b, c) for a in Fp for b in Fp for c in Fp]
    out += [(Fp(0), Fp(1), a, b) for a in Fp for b in Fp]
    out += [(Fp(0), Fp(0), Fp(1), a) for a in Fp]
    out += [(Fp(0), Fp(0), Fp(0), Fp(1))]
    return out

Xp = [P for P in P3points() if sum(t^4 for t in P) == 0]
print("#X(F_%d) = %d" % (p, len(Xp)))

from collections import defaultdict
fib4, fib2, undefined = defaultdict(list), defaultdict(list), 0
for P in Xp:
    vals, ok = {}, True
    for k, (n, d) in MAPS.items():
        dv = d(*P)
        if dv == 0:
            ok = False
            break
        vals[k] = n(*P)/dv
    if ok:
        fib4[(vals['w1'], vals['w2'], vals['y'], vals['z'])].append(P)
        fib2[(vals['w1'], vals['w2'])].append(P)
    else:
        undefined += 1

def dist(d):
    s = defaultdict(int)
    for v in d.values():
        s[len(v)] += 1
    return dict(sorted(s.items()))

print("points where some denominator vanishes :", undefined)
print("points where the map is defined        :", len(Xp) - undefined)
print()
print("fibre sizes of  P -> (w1,w2,y,z) :", dist(fib4), " <- injective, so the map is birational")
print("fibre sizes of  P -> (w1,w2)     :", dist(fib2))
print()
print("   The generic fibre has 4 geometric points, matching [k(K):k(w1,w2)] = 4. Fibres of size 2 are")
print("   those in which only two of the four are F_41-rational; the three large fibres lie over the")
print("   degenerate values of (w1,w2) where the two pencils fail to meet transversally.")
print("   Total accounted for:", sum(len(v) for v in fib2.values()))

#X(F_41) = 2520


points where some denominator vanishes : 1176
points where the map is defined        : 1344

fibre sizes of  P -> (w1,w2,y,z) : {1: 1344}  <- injective, so the map is birational
fibre sizes of  P -> (w1,w2)     : {2: 128, 4: 222, 64: 1, 66: 1, 70: 1}

   The generic fibre has 4 geometric points, matching [k(K):k(w1,w2)] = 4. Fibres of size 2 are
   those in which only two of the four are F_41-rational; the three large fibres lie over the
   degenerate values of (w1,w2) where the two pencils fail to meet transversally.
   Total accounted for: 1344


And an independent corroboration from the zeta function. If $X \cong \operatorname{Kum}(A)$ with
$A$ isogenous to a product of curves in the CM class of $j=1728$, and all $20$ Néron–Severi classes are
rational over $\mathbb{F}_p$ (they are, since $\mu_8 \subset \mathbb{F}_{41}$), then
$$\#X(\mathbb{F}_p) \;=\; p^2 + 1 + 20p + \bigl(\alpha^2+\bar\alpha^2\bigr),$$
where $\alpha = a+bi$ is the Frobenius eigenvalue of the CM curve and $p = a^2+b^2$, so that
$\alpha^2+\bar\alpha^2 = 2(a^2-b^2)$.

In [8]:
a, b = 5, 4
assert a^2 + b^2 == p
predicted = p^2 + 1 + 20*p + 2*(a^2 - b^2)
print("41 = %d^2 + %d^2,   alpha^2 + alphabar^2 = 2(%d - %d) = %d" % (a, b, a^2, b^2, 2*(a^2-b^2)))
print("predicted  p^2 + 1 + 20p + 2(a^2-b^2) =", predicted)
print("counted    #X(F_41)                   =", len(Xp))
print("agree                                 :", predicted == len(Xp))

41 = 5^2 + 4^2,   alpha^2 + alphabar^2 = 2(25 - 16) = 18
predicted  p^2 + 1 + 20p + 2(a^2-b^2) = 2520
counted    #X(F_41)                   = 2520
agree                                 : True


## 8. Points of $C\times C$ landing on the Fermat quartic

The last check runs the correspondence in the direction one actually wants to see. Take affine points
$(u_1,v_1)$, $(u_2,v_2)$ on $C$ over $\mathbb{F}_{41}$, form $(w_1,w_2,y,z)$ by the recipe of Section 3,
and look up the point of $X(\mathbb{F}_{41})$ with those coordinates.

Two things should happen. The fibres of $C\times C \to K$ should have size exactly $4$ — the orbits of
$\{\pm1\}\times\{(O,O),(T,T)\}$, where $-(u,v) = (u,-v)$ and $\tau(u,v) = (-u,-v)$. And the resulting
points should satisfy $x_0^4+x_1^4+x_2^4+x_3^4 = 0$.

In [9]:
half = Fp(1)/2

# lookup table: point of X  <->  its image (w1,w2,y,z)
table = {}
for P in Xp:
    vals, ok = {}, True
    for k, (n, d) in MAPS.items():
        dv = d(*P)
        if dv == 0:
            ok = False
            break
        vals[k] = n(*P)/dv
    if ok:
        table[(vals['w1'], vals['w2'], vals['y'], vals['z'])] = P

Cpts = [(u, v) for u in Fp for v in Fp if v^2 == (u^2 - 1)*(u^2 - half)]
print("#C(F_41), affine points:", len(Cpts))

fib = defaultdict(list)
for (u1, v1) in Cpts:
    for (u2, v2) in Cpts:
        w1, w2 = u1^2, u2^2
        if v2 == 0 or u1 == 0 or w1 == half:
            continue
        fib[(w1, w2, v1*(w2-1)/(v2*(w1-half)), u2/u1)].append(((u1, v1), (u2, v2)))

print("fibre sizes of  C x C -> K :", dist(fib))

def orbit(P, Q):
    neg = lambda R: (R[0], -R[1])
    tau = lambda R: (-R[0], -R[1])
    return sorted({(P, Q), (neg(P), neg(Q)), (tau(P), tau(Q)), (neg(tau(P)), neg(tau(Q)))})

ok = all(sorted(v) == orbit(*v[0]) for v in fib.values())
print("every fibre is exactly the orbit of {+-1} x {(O,O),(T,T)} :", ok)
print()

hits = [(k, table[k], fib[k]) for k in fib if k in table]
print("keys reached from C x C that are F_41-rational on X:", len(hits))
for k, P, src in sorted(hits)[:3]:
    print()
    print("   (w1,w2,y,z) =", k)
    print("   preimages in C x C :", src)
    print("   point of X         :", P, "    x0^4+x1^4+x2^4+x3^4 =", sum(t^4 for t in P))

#C(F_41), affine points: 30
fibre sizes of  C x C -> K : {4: 169}
every fibre is exactly the orbit of {+-1} x {(O,O),(T,T)} : True

keys reached from C x C that are F_41-rational on X: 92

   (w1,w2,y,z) = (4, 4, 5, 1)
   preimages in C x C : [((2, 20), (2, 21)), ((2, 21), (2, 20)), ((39, 20), (39, 21)), ((39, 21), (39, 20))]
   point of X         : (1, 35, 23, 27)     x0^4+x1^4+x2^4+x3^4 = 0

   (w1,w2,y,z) = (4, 4, 5, 40)
   preimages in C x C : [((2, 20), (39, 21)), ((2, 21), (39, 20)), ((39, 20), (2, 21)), ((39, 21), (2, 20))]
   point of X         : (1, 7, 16, 3)     x0^4+x1^4+x2^4+x3^4 = 0

   (w1,w2,y,z) = (4, 4, 36, 1)
   preimages in C x C : [((2, 20), (2, 20)), ((2, 21), (2, 21)), ((39, 20), (39, 20)), ((39, 21), (39, 21))]
   point of X         : (1, 7, 25, 38)     x0^4+x1^4+x2^4+x3^4 = 0


## 9. What is verified, and what is not

Verified above, all of it machine-checked:

- $C$ has $j = 287496$ and CM by $\mathbb{Z}[2i]$; $C' = C/\tau$ has $j=1728$ and CM by $\mathbb{Z}[i]$.
- $A \cong C'\times C \cong E_i\times E_{2i}$, hence $A \not\cong E\times E$, only $2$-isogenous to it.
- The identities (7) hold modulo $F$ and not identically.
- Swinnerton-Dyer's $F_2/F_1$ and $G_2/G_1$ collapse to $\Theta\, h'_{13}/e'_+$ and $\Theta\, h'_{31}/e'_-$,
  by exact cancellation.
- Both equations (6) of $K$ pull back to zero modulo $F$ — exact ideal membership over
  $\mathbb{Q}(\zeta_8)$, bidegree $8/8$ throughout.
- Over $\mathbb{F}_{41}$ the map is injective on all points where it is defined, hence birational, hence
  (both surfaces being K3) an isomorphism; the map to $(w_1,w_2)$ is generically $4:1$.
- $\#X(\mathbb{F}_{41}) = 2520$ matches the Kummer prediction exactly.
- The fibres of $C\times C\to K$ are precisely the orbits of $\{\pm1\}\times\{(O,O),(T,T)\}$.

Not done here: the **inverse map** $C\times C \dashrightarrow \mathbb{P}^3$ in closed form. Recovering
$(x_0:x_1:x_2:x_3)$ as rational functions of $(w_1,w_2,y,z)$ means inverting a bidegree-$8/8$ system,
i.e. an elimination that has not been attempted. Everything above shows the inverse exists and is
defined over $\mathbb{Q}(\mu_8)$; it does not write it down.

## References

- E. Ieronymou, A. N. Skorobogatov, Yu. G. Zarhin, *On the Brauer group of diagonal quartic surfaces*,
  with an appendix by Sir Peter Swinnerton-Dyer, J. London Math. Soc. **83** (2011) 659–672,
  [arXiv:0912.2865](https://arxiv.org/abs/0912.2865). Theorem A.1 and the appendix are the source of
  everything in Sections 3–6.
- M. Mizukami, *Birational mappings from quartic surfaces to Kummer surfaces*, Master's thesis,
  University of Tokyo, 1977 (in Japanese) — the original.
- I. I. Piatetskii-Shapiro, I. R. Shafarevich, *Torelli's theorem for algebraic surfaces of type K3*,
  Izv. Akad. Nauk SSSR **35** (1971) 530–572 — where the isomorphism was first observed to exist,
  non-constructively.
- T. Shioda, N. Mitani, *Singular abelian surfaces and binary quadratic forms*, in: Classification of
  Algebraic Varieties and Compact Complex Manifolds, LNM **412** (1974) 259–287 — the
  form/surface dictionary used in Section 2.